### PR TITLE
feat(tap): Add duration flag to vector tap

### PR DESCRIPTION
**Pull request by ArunPiduguDD**
### Overview

Adds optional `duration` flag to `vector tap` command, which allows specifying how long to run `tap` before exiting (by default `tap` will run indefinitely)

### Testing

Tested by using `tap` on a `demo_logs` test pipeline (both with and without the duration flag)

### CODE REVIEW

File: src/tap/cmd.rs
Comment (bruceg) [inline src/tap/cmd.rs:117]: Both of these calls to `elapsed` end up invoking a system call. While it is very fast, we should probably cache that result, something like:
```suggestion
        let elapsed = start_time.elapsed();
        if elapsed >= stream_duration {
            return exitcode::OK;
        }

        let message = timeout(stream_duration - elapsed, stream.next()).await;
```

File: src/tap/cmd.rs
Comment (bruceg) [inline src/tap/cmd.rs:106]: nit: consider this may be simpler:
```suggestion
    let stream_duration = opts.duration.map(Duration::from_millis).unwrap_or(Duration::MAX);
```

File: src/tap/cmd.rs
Comment (bruceg) [inline src/tap/cmd.rs:122]: Given this pattern, we should probably use `match message` to encompass all three of the cases (including the `else` at the end of the function).

File: src/tap/cmd.rs
Comment (bruceg) [inline src/tap/cmd.rs:119]: Is this right, that we just return immediately with no warning if the message is in error?

File: src/tap/cmd.rs
Comment (ArunPiduguDD) [reply src/tap/cmd.rs:106]: Sounds good, will update with these changes

Comment (ArunPiduguDD) [reply src/tap/cmd.rs:106]: Yea that would work better, though going with @jszwedko's suggestion to remove this entirely

File: src/tap/cmd.rs
Comment (ArunPiduguDD) [reply src/tap/cmd.rs:119]: Yea, idea was that this will timeout if it reaches the duration specified by the user, to handle a case where the current interval lasts longer than the duration (an extreme example would be that the duration is 5 seconds but the interval is 1 minute), in which case the time `tap` would be running for 1 minute before exiting even though technically the duration is less.

The other option would be to just leave only the `if start_time.elapsed() >= stream_duration` check, it depends on whether we think the case I described above is something we want to cover.

File: src/tap/cmd.rs
Comment (ArunPiduguDD) [reply src/tap/cmd.rs:106]: @jszwedko I think if we go with the current timeout approach, then leaving `stream_duration` as `Option` would become messy (we have to add more logic to call `stream.next()` both with or without the timeout, as well as process the results for each call slightly differently)

File: src/tap/mod.rs
Comment (jszwedko) [inline src/tap/mod.rs:58]: Could we call this `duration_ms`? This is per https://github.com/vectordotdev/vector/blob/master/docs/specs/configuration.md#naming

**Review by jszwedko — COMMENTED** [review]
One last comment from me, otherwise I think this looks good. I agree that having the `stream_duration` be an `Option` would complicate the later checks so I'm ok leaving it as-is even if the modeling is a bit indirect.

File: src/tap/mod.rs
Comment (ArunPiduguDD) [reply src/tap/mod.rs]: Sounds good, just updated

**Review by jszwedko — APPROVED** [review]
Thanks! I'll let @bruceg give this another look, as well as the documentation team, before we merge this.

**Review by pront — APPROVED** [review]
👍 Thank you!

**Review by estherk15 — APPROVED** [review]

File: changelog.d/vector_tap_duration_flag.enhancement.md
Comment (jszwedko) [inline changelog.d/vector_tap_duration_flag.enhancement.md:3]:
```suggestion
automatically stop after that duration has elapsed.

authors: ArunPiduguDD
```

I forgot this is required for contributions.

**Review by bruceg — APPROVED** [review]

**Pull request merged**
